Create rule S6457 (#1261)
SonarSource/rspec: rules/S6457/java/rule.adoc

 *public static void main(String... args) {
 *    System.out.println("Hello World!");
 *}}
 */
public void f1() {}

/**
 * {@snippet name="snippet2" :
 *public static void main(String... args) {
 *    System.out.println("Hello World!");  // @highlight substring="out"
 *}}
 */
public void f2() {}

/**
 * {@snippet name="snippet3" :
 *public static void main(String... args) {
 *    System.out.println("Hello World!");  // @highlight substring="System"
 *}}
 */
public void f3() {}
----

== Compliant Solution

This snippet is external 

[source,java]
----
/**
 * {@snippet name="snippet1" file="HelloWorld.java"}
 */
public void f1() {}
----

The following 2 snippets are hybrid so `@highlight` can be used here

[source,java]
----
/**
 * {@snippet name="snippet2" file="HelloWorld.java":
 * public static void main(String... args) {
 *     System.out.println("Hello World!");  // @highlight substring="out"
 * }}
 */
public void f2() {}

/**
 * {@snippet name="snippet3" file="HelloWorld.java":
 * public static void main(String... args) {
 *     System.out.println("Hello World!");  // @highlight substring="System"
 * }}
 */
public void f3() {}
----

> And the content of the HelloWorld.java is:

Reviewer: ```suggestion
And the content of HelloWorld.java is:
```

> Code snippets are useful for adding code examples to javadoc comments. This will help developers to better understand the usage of documented method. However. Sometimes you can have the same code example duplicated in different inline snippets. In this case you can benefit from having an external or a hybrid snippet by extracting the code example into a separate file.

Reviewer: ```suggestion
Code snippets are useful for adding code examples to javadoc comments. This will help developers to better understand the usage of documented methods. However, sometimes you can have the same code example duplicated in different inline snippets. In this case you can benefit from having an external or a hybrid snippet by extracting the code example into a separate file.
```

== Noncompliant Code Example

> With the default threshold of 3:

Reviewer: Do you think it is worth explaining a bit more what this is?

[source,java]
----
/**
 * {@snippet name="snippet1" :            // Noncompliant, this snippet is duplicated
 *public static void main(String... args) {
 *    System.out.println("Hello World!");
 *}}
 */
public void f1() {}

/**
 * {@snippet name="snippet2" :
 *public static void main(String... args) {
 *    System.out.println("Hello World!");  // @highlight substring="out"
 *}}
 */
public void f2() {}

/**

>  * {@snippet name="snippe3" :

Reviewer: ```suggestion
 * {@snippet name="snippet3" :
```

 *public static void main(String... args) {
 *    System.out.println("Hello World!");  // @highlight substring="System"
 *}}
 */
public void f3() {}
----

== Compliant Solution

[source,java]
----
/**
 * {@snippet name="snippet1" file="HelloWorld.java"}
 */
public void f1() {}

/**
 * {@snippet name="snippet2" file="HelloWorld.java":
 * public static void main(String... args) {
 *     System.out.println("Hello World!");  // @highlight substring="out"
 * }}
 */
public void f2() {}

/**
 * {@snippet name="snippet3" file="HelloWorld.java":
 * public static void main(String... args) {
 *     System.out.println("Hello World!");  // @highlight substring="System"
 * }}
 */

> public void f3() {}

Reviewer: A comment about the fact that they are hybrid snippets would be very helpful in these snippets.

----

== See

Oracle’s Programmer's Guide to Snippets: https://docs.oracle.com/en/java/javase/18/code-snippet/index.html[https://docs.oracle.com/en/java/javase/18/code-snippet/index.html]

ifdef::env-github,rspecator-view[]

'''
== Implementation Specification
(visible only on this page)

=== Message

> Extract the content of the @nippet to the separate file instead of duplicating it {number} times.

Reviewer: ```suggestion
Extract the content of the @snippet to a separate file instead of duplicating it {number} times.
```
